src/Framework/Framework/Compilation/ViewCompiler/DefaultViewCompilerCodeEmitter.cs

```diff
@@ -0,0 +1,360 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using DotVVM.Framework.Binding;
+using System.Collections.Concurrent;
+using System.Linq.Expressions;
+using Microsoft.Extensions.DependencyInjection;
+using DotVVM.Framework.Controls;
+
+namespace DotVVM.Framework.Compilation
+{
+    public class DefaultViewCompilerCodeEmitter
+    {
+        private static Type[] emptyTypeArguments = new Type[] { };
+
+        private int CurrentControlIndex;
+
+        public const string ControlBuilderFactoryParameterName = "controlBuilderFactory";
+        public const string ServiceProviderParameterName = "services";
+        public const string BuildTemplateFunctionName = "BuildTemplate";
+
+        private Dictionary<GroupedDotvvmProperty, string> cachedGroupedDotvvmProperties = new Dictionary<GroupedDotvvmProperty, string>();
+        private ConcurrentDictionary<(Type obj, string argTypes), string> injectionFactoryCache = new ConcurrentDictionary<(Type obj, string argTypes), string>();
+        private readonly Stack<BlockInfo> blockStack = new();
```

Comment:
```suggestion
        private readonly Stack<BlockInfo> blockStack = new();
        private currentBlock => blockStack.Peek()
```

I think this could improve the code a bit.